### PR TITLE
CORTX-29142: Update the algorithm for calculating parity buffers.

### PR DESCRIPTION
**Pull request by Shreya-18**
Updated formula calculates the percentage of parity buffers from the
total bytes received from motr with help of N and K and then discards
it from total bytecount.
This gives us more accurate results than previous computation.

### Testing
Config : 4+2
Written data of 9mb
```
root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ ./test2/cortx-motr/motr/st/utils/m0cp -l inet:tcp:192.168.56.39@21501 -H inet:tcp:192.168.56.39@22001 -p 0x7000000000000001:0xe9 -P 0x7           200000000000001:0x2b -o 20:25 -s 1m -c 9 -L 9 -b 10 src_file
```

results are of 10mb
```                                                                        
root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ consul kv get -recurse | grep ioser
ioservices/0x7200000000000001:0x1c/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 2097152, "object_cnt": 1}
ioservices/0x7200000000000001:0x3b/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 2097152, "object_cnt": 1}
ioservices/0x7200000000000001:0x4a/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 3145728, "object_cnt": 1}
ioservices/0x7200000000000001:0x69/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 3145728, "object_cnt": 1}
ioservices/0x7200000000000001:0x78/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 2097152, "object_cnt": 1}
ioservices/0x7200000000000001:0xd/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 3145728, "object_cnt": 1}

root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ consul kv get -recurse | grep bytec
bytecount/critical:0
bytecount/damaged:0
bytecount/degraded:0
bytecount/healthy:10485760

root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ hctl status
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 10485760
Data pool:
    # fid name
    0x6f00000000000001:0x99 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xe9 'default': 'the pool' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:192.168.56.39@22001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:192.168.56.39@21001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:192.168.56.39@21002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:192.168.56.39@21003
    [offline]  m0_client_other  0x7200000000000001:0x2b  inet:tcp:192.168.56.39@21501
    [unknown]  m0_client_other  0x7200000000000001:0x2e  inet:tcp:192.168.56.39@21502
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:192.168.51.181@22001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:192.168.51.181@21001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:192.168.51.181@21002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:192.168.51.181@21003
    [unknown]  m0_client_other  0x7200000000000001:0x59  inet:tcp:192.168.51.181@21501
    [unknown]  m0_client_other  0x7200000000000001:0x5c  inet:tcp:192.168.51.181@21502
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [started]  hax        0x7200000000000001:0x63  inet:tcp:192.168.47.173@22001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:192.168.47.173@21001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:192.168.47.173@21002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:192.168.47.173@21003
    [unknown]  m0_client_other  0x7200000000000001:0x87  inet:tcp:192.168.47.173@21501
    [unknown]  m0_client_other  0x7200000000000001:0x8a  inet:tcp:192.168.47.173@21502
```

for writing 4mb, we get 4mb in hctl status

```
root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ ./test2/cortx-motr/motr/st/utils/m0cp -l inet:tcp:192.168.56.39@21501 -H inet:tcp:192.168.56.39@22001 -p 0x7000000000000001:0xe9 -P 0x7200000000000001:0x2b -o 20:21 -s 1m -c 4 -L 9 -b 10 src_file

root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ consul kv get -recurse | grep ioser
ioservices/0x7200000000000001:0x1c/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}
ioservices/0x7200000000000001:0x3b/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}
ioservices/0x7200000000000001:0x4a/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}
ioservices/0x7200000000000001:0x69/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}
ioservices/0x7200000000000001:0x78/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}
ioservices/0x7200000000000001:0xd/pvers/0x7600000000000001:0x9a/users/8881212:{"bc": 1048576, "object_cnt": 1}

$ journalctl -xe | grep "parity"
Mar 08 09:07:27 ssc-vm-g3-rhev4-2743.colo.seagate.com hare-hax[106057]: 2022-03-08 09:07:27,191 [DEBUG] {byte-count-updater} Received pool version and status: {'0x7600000000000001:0x9a': PverInfo(fid=0x7600000000000001:0x9a, state=0, data_units=4, parity_units=2, pool_width=18, unit_size=0)}
Mar 08 09:07:27 ssc-vm-g3-rhev4-2743.colo.seagate.com hare-hax[106057]: 2022-03-08 09:07:27,197 [DEBUG] {byte-count-updater} Bytecount with parity buffer: {'0x7600000000000001:0x9a': 6291456}
Mar 08 09:07:27 ssc-vm-g3-rhev4-2743.colo.seagate.com hare-hax[106057]: 2022-03-08 09:07:27,197 [DEBUG] {byte-count-updater} Bytecount without parity buffer: {'0x7600000000000001:0x9a': 4194304}

root@ssc-vm-g3-rhev4-2743.colo.seagate.com:/root
$ hctl status
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 4194304
Data pool:
    # fid name
    0x6f00000000000001:0x99 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xe9 'default': 'the pool' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:192.168.56.39@22001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:192.168.56.39@21001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:192.168.56.39@21002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:192.168.56.39@21003
    [offline]  m0_client_other  0x7200000000000001:0x2b  inet:tcp:192.168.56.39@21501
    [unknown]  m0_client_other  0x7200000000000001:0x2e  inet:tcp:192.168.56.39@21502
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:192.168.51.181@22001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:192.168.51.181@21001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:192.168.51.181@21002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:192.168.51.181@21003
    [unknown]  m0_client_other  0x7200000000000001:0x59  inet:tcp:192.168.51.181@21501
    [unknown]  m0_client_other  0x7200000000000001:0x5c  inet:tcp:192.168.51.181@21502
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [started]  hax        0x7200000000000001:0x63  inet:tcp:192.168.47.173@22001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:192.168.47.173@21001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:192.168.47.173@21002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:192.168.47.173@21003
    [unknown]  m0_client_other  0x7200000000000001:0x87  inet:tcp:192.168.47.173@21501
    [unknown]  m0_client_other  0x7200000000000001:0x8a  inet:tcp:192.168.47.173@21502
```

### CODE REVIEW

File: hax/hax/bytecount.py [hunk 1]
```diff
@@ -113,8 +113,9 @@ def _get_parity_buffers(self, bc: int, state: PverInfo) -> int:
         Calculate the parity buffer count based on pool configuration.
         """
         tot_units = state.data_units + state.parity_units
-        bytes_per_unit = ceil(bc / tot_units)
-        return bytes_per_unit * state.parity_units
+        percent = state.parity_units / tot_units
+        LOG.debug('Parity buffer percentage %s', percent)
+        return ceil(bc * percent)
```
Comment (madhavemuri) [inline hax/hax/bytecount.py:118]: Is ceiling is a must here ?

Comment (Shreya-18) [reply hax/hax/bytecount.py:118]: while calculating parity percentage, we get results in floats.
```
$ hctl status
Bytecount:
critical : 0
damaged : 0
degraded : 0
healthy : 34952533.333333336
```

Confirmed with @yeshpal-jain-seagate  and @AbhishekSahaSeagate for using `ceil` here. Let me know, if any change is required.

Comment (madhavemuri) [reply hax/hax/bytecount.py:118]: okay, fine.

**Review by madhavemuri — APPROVED** [review]

**Comment by vaibhavparatwar** [comment]
@mssawant @SwapnilGaonkar7  need your help to review this so that we can get this in to feature branch. We are targetting to merge dg_bytecount branch to main on Monday or Tuesday.

File: hax/hax/bytecount.py (hunk 1 above)
Comment (mssawant) [reply hax/hax/bytecount.py:118]: ceiling is required as the result may be a fraction.

Comment (mssawant) [reply hax/hax/bytecount.py:118]: @madhavemuri, why do we need percentage here?
So the earlier algorithm would work the same way, consider the following example with partially populated parity group, with unit size of 1M, N = 4, K = 2 and data written = 9M
we have 9M of data and 6M of parity = 15M of total bytecount, now we do,

scale down the entire bytecount and virtaully concentrate it in a single parity group, ceil(15/6) = 3M (virtual temporary unit size)
Now we do, K * 3M = 6M
actual bytecount = 15M - 6M = 9M
So the above calculate will work for both partially populated and fully populated parity groups, how would using percentage help to do it any better? cc @Shreya-18

**Comment by Shreya-18** [comment]
> @Shreya-18, also looks like your local `dg_bytecount` branch is not updated with remote as I see multiple commits.

yes, rebased. Thanks!

**Comment by supriyachavan4398** [comment]
Created custom build with new changes: https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5290/
Did 3N deployment with above build: https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/789/
```
[root@ssc-vm-g2-rhev4-2610 ~]# kubectl get pods
NAME                                                 READY   STATUS    RESTARTS   AGE
consul-client-bpr9h                                  1/1     Running   0          16m
consul-client-j25xh                                  1/1     Running   0          16m
consul-client-w99lg                                  1/1     Running   0          16m
consul-server-0                                      1/1     Running   0          15m
consul-server-1                                      1/1     Running   0          16m
consul-server-2                                      1/1     Running   0          16m
cortx-control-868bf5d766-fv48n                       1/1     Running   0          9m25s
cortx-data-ssc-vm-g2-rhev4-2610-7df95477cb-fgbf6     4/4     Running   0          8m26s
cortx-data-ssc-vm-g2-rhev4-2611-769f4b947-mgncz      4/4     Running   0          8m25s
cortx-data-ssc-vm-g2-rhev4-2621-859f4c6ffb-mz4z8     4/4     Running   0          8m24s
cortx-ha-74678f5bd4-lf7pv                            3/3     Running   0          5m33s
cortx-server-ssc-vm-g2-rhev4-2610-8f6cb857d-sqdm8    2/2     Running   0          6m40s
cortx-server-ssc-vm-g2-rhev4-2611-7bdf9c7c4-jbn7k    2/2     Running   0          6m39s
cortx-server-ssc-vm-g2-rhev4-2621-5458f6459d-jpjwb   2/2     Running   0          6m38s
kafka-0                                              1/1     Running   0          12m
kafka-1                                              1/1     Running   0          12m
kafka-2                                              1/1     Running   0          12m
openldap-0                                           1/1     Running   0          16m
openldap-1                                           1/1     Running   0          16m
openldap-2                                           1/1     Running   0          15m
zookeeper-0                                          1/1     Running   0          14m
zookeeper-1                                          1/1     Running   0          14m
zookeeper-2                                          1/1     Running   0          14m
[root@ssc-vm-g2-rhev4-2610 ~]#
```
```
[root@ssc-vm-g2-rhev4-2610 ~]# kubectl exec -it cortx-data-ssc-vm-g2-rhev4-2610-7df95477cb-fgbf6 -c cortx-hax -- /bin/bash
[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-2610 /]# hctl status
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_COLLATE to default locale: No such file or directory
Bytecount:
    critical : 0
    damaged : 0
    degraded : 0
    healthy : 0
Data pool:
    # fid name
    0x6f00000000000001:0x93 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xd2 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2610
    [started]  hax                 0x7200000000000001:0x2b         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@22001
    [started]  ioservice           0x7200000000000001:0x2e         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21001
    [started]  ioservice           0x7200000000000001:0x3b         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21002
    [started]  confd               0x7200000000000001:0x48         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21003
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2621
    [started]  hax                 0x7200000000000001:0x7          inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@22001
    [started]  ioservice           0x7200000000000001:0xa          inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21001
    [started]  ioservice           0x7200000000000001:0x17         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21002
    [started]  confd               0x7200000000000001:0x24         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21003
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2611  (RC)
    [started]  hax                 0x7200000000000001:0x4f         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@22001
    [started]  ioservice           0x7200000000000001:0x52         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21001
    [started]  ioservice           0x7200000000000001:0x5f         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21002
    [started]  confd               0x7200000000000001:0x6c         inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21003
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2621
    [started]  hax                 0x7200000000000001:0x71         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2621@22001
    [started]  rgw                 0x7200000000000001:0x74         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2621@21501
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2611
    [started]  hax                 0x7200000000000001:0x79         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2611@22001
    [started]  rgw                 0x7200000000000001:0x7c         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2611@21501
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2610
    [started]  hax                 0x7200000000000001:0x81         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2610@22001
    [started]  rgw                 0x7200000000000001:0x84         inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2610@21501
[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-2610 /]#
```
cc. @Shreya-18, @mssawant, @vaibhavparatwar

**Comment by vaibhavparatwar** [comment]
@mssawant @SwapnilGaonkar7 please review this PR as you get chance.. Need this to be merged to feature branch of dg_bytecount. cc @Shreya-18

**Comment by SwapnilGaonkar7** [comment]
@Shreya-18 Have we tested this changes different combinations of data, parity, spare to see if we are getting expected result?

**Comment by Shreya-18** [comment]
> @Shreya-18 Have we tested this changes different combinations of data, parity, spare to see if we are getting expected result?

Yes, in fact, this was tested along with motr guys while integration testing.

File: hax/hax/bytecount.py (hunk 1 above)
Comment (yeshpal-jain-seagate) [reply hax/hax/bytecount.py:118]: how do we know what is the unit size for that stripe, we do not maintain any unit size parameter for a particular size.

Comment (mssawant) [reply hax/hax/bytecount.py:118]: @yeshpal-jain-seagate, why do we need to know the unit size? Here we are reducing multiple parity groups to a single group and extending the unit size to a new virtual unit size as we just need to separate out the parity from actual data.

**Comment by mssawant** [comment]
@Shreya-18, I don't the need for this patch anymore. Let's close the PR. cc @vaibhavparatwar, @madhavemuri.

**Comment by vaibhavparatwar** [comment]
Closing PR as per @mssawant 's comment

**Pull request closed without merging**